### DISCUSSION TITLE
How to create github action secrets

### DISCUSSION THREAD

**Discussion by tz-torchai** [pulumi/pulumi#9377] [Pulumi Providers] [answered]
In the API doc, we can pass to `ActionsSecret` a github action secret that's encrypted by github public key:

https://www.pulumi.com/registry/packages/github/api-docs/actionssecret/
```python
import pulumi
import pulumi_github as github

example_public_key = github.get_actions_public_key(repository="example_repository")
example_secret_actions_secret = github.ActionsSecret("exampleSecretActionsSecret",
    repository="example_repository",
    secret_name="example_secret_name",
    plaintext_value=var["some_secret_string"])
example_secret_index_actions_secret_actions_secret = github.ActionsSecret("exampleSecretIndex/actionsSecretActionsSecret",
    repository="example_repository",
    secret_name="example_secret_name",
    encrypted_value=var["some_encrypted_secret_string"])
```

I am using Python and I am wondering

1. how to create this encrypted secret `some_encrypted_secret_string`? It's said to be encrypted using the github public key, but how, which algorithm/library to use?
2. what does this github public key come from, how to get it set up for a repository?

**Comment by nimbinatus** [comment, marked as answer]
To create the encrypted secret, you'll use the encryption method of your language of choice. I know it's supposed to be base64 encoded, so Python has the built-in base64 library for that part. It looks like GitHub recommends using LibSodium [based on the information in their API docs](https://docs.github.com/en/rest/reference/actions#create-or-update-a-repository-secret--parameters). They've got an [example putting those two libraries together to build the encoded secret](https://docs.github.com/en/rest/reference/actions#example-encrypting-a-secret-using-python) in their docs.

Regarding getting the key, I'd probably use the [`getActionsPublicKey` call](https://www.pulumi.com/registry/packages/github/api-docs/getactionspublickey/). I'm pretty sure there's other ways to do it, but I'd stay in Pulumi to get it.

Let me know if this helped!

**Comment by phitoduck** [comment]
@tz-torchai @nimbinatus were you ever able to get this to work?

I tried the solution in the selected answer, but it appeared that the `[getActionsPublicKey call](https://www.pulumi.com/registry/packages/github/api-docs/getactionspublickey/)` gave me a `404 Error` when I ran `pulumi up`. I created an issue for it here:

https://github.com/pulumi/pulumi-github/issues/248#issue-1393340785
